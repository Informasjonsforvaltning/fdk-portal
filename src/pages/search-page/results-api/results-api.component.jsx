import React from 'react';
import PropTypes from 'prop-types';
import ReactPaginate from 'react-paginate';
import { Button, Modal, ModalBody, ModalFooter, ModalHeader } from 'reactstrap';
import cx from 'classnames';
import _ from 'lodash';
import { withRouter } from 'react-router';

import localization from '../../../lib/localization';
import { SearchHitItem } from './search-hit-item/search-hit-item.component';
import { FilterBox } from '../../../components/filter-box/filter-box.component';
import { FilterTree } from '../filter-tree/filter-tree.component';
import { getSortfield, setPage, setSortfield } from '../search-location-helper';
import { parseSearchParams } from '../../../lib/location-history-helper';
import { FilterPills } from '../filter-pills/filter-pills.component';

const renderFilterModal = ({
  showFilterModal,
  closeFilterModal,
  apiAggregations,
  locationSearch,
  publisherCounts,
  publishers,
  onFilterFormat,
  onFilterPublisherHierarchy
}) => (
  <Modal isOpen={showFilterModal} toggle={closeFilterModal}>
    <ModalHeader toggle={closeFilterModal}>{localization.filter}</ModalHeader>
    <ModalBody>
      <div className="search-filters">
        <FilterTree
          title={localization.provider}
          aggregations={publisherCounts}
          handleFiltering={onFilterPublisherHierarchy}
          activeFilter={locationSearch.orgPath}
          referenceDataItems={publishers}
        />
        <FilterBox
          htmlKey={2}
          title={localization.facet.format}
          filter={_.get(apiAggregations, 'formats')}
          onClick={onFilterFormat}
          activeFilter={locationSearch.format}
        />
      </div>
    </ModalBody>
    <ModalFooter>
      <Button className="fdk-button" onClick={closeFilterModal} color="primary">
        {localization.close}
      </Button>
    </ModalFooter>
  </Modal>
);

const renderHits = (hits, publishers, referenceData) => {
  if (hits && Array.isArray(hits)) {
    return hits.map((item, index) => (
      <SearchHitItem
        key={item.id}
        item={item}
        fadeInCounter={index < 3 ? index : null}
        publishers={publishers}
        referenceData={referenceData}
      />
    ));
  }
  return null;
};

export const ResultsApiPure = ({
  showFilterModal,
  closeFilterModal,
  apiItems,
  apiTotal,
  apiAggregations,
  onFilterPublisherHierarchy,
  onFilterFormat,
  publisherCounts,
  publishers,
  hitsPerPage,
  history,
  location,
  referenceData
}) => {
  const locationSearch = parseSearchParams(location);

  const page = parseInt(locationSearch.page || 0, 10);
  const pageCount = Math.ceil((apiTotal || 1) / hitsPerPage);

  const sortfield = getSortfield(location);
  const sortByScoreClass = cx('fdk-button', 'fdk-button-black-toggle', {
    selected: !sortfield
  });
  const sortByLastModifiedClass = cx('fdk-button', 'fdk-button-black-toggle', {
    selected: sortfield === 'modified'
  });

  const onSortByScoreClick = () => {
    setSortfield(history, location, undefined);
  };
  const onSortByModifiedClick = () => {
    setSortfield(history, location, 'modified');
  };

  const onPageChange = data => {
    setPage(history, location, data.selected);
    window.scrollTo(0, 0);
  };

  return (
    <main data-test-id="apis" id="content">
      <div className="row mb-3">
        <div className="col-6 col-lg-4" />
        <div className="col-6 col-lg-4 offset-lg-4">
          <div className="d-flex justify-content-end">
            <Button
              className={sortByScoreClass}
              onClick={onSortByScoreClick}
              color="primary"
            >
              {localization.sort.relevance}
            </Button>
            <Button
              className={sortByLastModifiedClass}
              onClick={onSortByModifiedClick}
              color="primary"
            >
              {localization.sort.modified}
            </Button>
          </div>
        </div>
      </div>
      <div className="row">
        <aside className="search-filters col-lg-4 d-none d-lg-block">
          <span className="uu-invisible" aria-hidden="false">
            Filtrering
          </span>

          <FilterPills
            history={history}
            location={location}
            locationSearch={locationSearch}
            publishers={publishers}
          />

          {apiAggregations && (
            <div>
              {renderFilterModal({
                showFilterModal,
                closeFilterModal,
                apiAggregations,
                locationSearch,
                publisherCounts,
                publishers,
                onFilterFormat,
                onFilterPublisherHierarchy
              })}
              <FilterTree
                title={localization.provider}
                aggregations={publisherCounts}
                handleFiltering={onFilterPublisherHierarchy}
                activeFilter={locationSearch.orgPath}
                referenceDataItems={publishers}
              />
              <FilterBox
                htmlKey={2}
                title={localization.facet.format}
                filter={_.get(apiAggregations, 'formats')}
                onClick={onFilterFormat}
                activeFilter={locationSearch.format}
              />
            </div>
          )}
        </aside>
        <div id="apis" className="col-12 col-lg-8">
          {renderHits(apiItems, publishers, referenceData)}

          <div className="col-12 d-flex justify-content-center">
            <span className="uu-invisible" aria-hidden="false">
              Sidepaginering.
            </span>
            <ReactPaginate
              pageCount={pageCount}
              pageRangeDisplayed={2}
              marginPagesDisplayed={1}
              previousLabel={localization.page.prev}
              nextLabel={localization.page.next}
              breakLabel={<span>...</span>}
              breakClassName="break-me"
              containerClassName="pagination"
              onPageChange={onPageChange}
              subContainerClassName="pages pagination"
              activeClassName="active"
              forcePage={page}
              disableInitialCallback
            />
          </div>
        </div>
      </div>
    </main>
  );
};

ResultsApiPure.defaultProps = {
  showFilterModal: false,
  closeFilterModal: _.noop,

  apiItems: [],
  apiTotal: 0,
  apiAggregations: null,

  onFilterPublisherHierarchy: _.noop,
  onFilterFormat: _.noop,
  publisherCounts: null,
  publishers: null,

  hitsPerPage: 10,

  history: { push: _.noop },
  location: { search: '' },
  referenceData: null
};

ResultsApiPure.propTypes = {
  showFilterModal: PropTypes.bool,
  closeFilterModal: PropTypes.func,

  apiItems: PropTypes.array,
  apiTotal: PropTypes.number,
  apiAggregations: PropTypes.object,

  onFilterPublisherHierarchy: PropTypes.func,
  onFilterFormat: PropTypes.func,

  publisherCounts: PropTypes.array,
  publishers: PropTypes.object,

  hitsPerPage: PropTypes.number,

  history: PropTypes.object,
  location: PropTypes.object,
  referenceData: PropTypes.object
};

export const ResultsApi = withRouter(ResultsApiPure);
