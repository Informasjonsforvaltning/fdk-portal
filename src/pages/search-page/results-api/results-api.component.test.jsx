import React from 'react';
import { shallow } from 'enzyme';
import _ from 'lodash';
import { ResultsApiPure } from './results-api.component';
import apisApiResponse from '../../../mock/apis.response.json';

test('should render ResultsApi correctly with minimum of props and no hits', () => {
  const result = shallow(<ResultsApiPure />);
  expect(result).toMatchSnapshot();
});

test('should render ResultsApi correctly with hits', () => {
  const props = {
    apiItems: _.get(apisApiResponse, 'hits'),
    apiAggregations: _.get(apisApiResponse, 'aggregations'),
    apiTotal: _.get(apisApiResponse, 'total')
  };
  const result = shallow(<ResultsApiPure {...props} />);
  expect(result).toMatchSnapshot();
});
