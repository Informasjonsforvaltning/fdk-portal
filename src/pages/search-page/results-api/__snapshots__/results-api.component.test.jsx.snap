// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ResultsApi correctly with hits 1`] = `
<main
  data-test-id="apis"
  id="content"
>
  <div
    className="row mb-3"
  >
    <div
      className="col-6 col-lg-4"
    />
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <aside
      className="search-filters col-lg-4 d-none d-lg-block"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Filtrering
      </span>
      <FilterPills
        history={
          Object {
            "push": [Function],
          }
        }
        location={
          Object {
            "search": "",
          }
        }
        locationSearch={Object {}}
        losItems={null}
        publishers={null}
        themesItems={null}
      />
      <div>
        <Modal
          autoFocus={true}
          backdrop={true}
          backdropTransition={
            Object {
              "mountOnEnter": true,
              "timeout": 150,
            }
          }
          centered={false}
          fade={true}
          isOpen={false}
          keyboard={true}
          modalTransition={
            Object {
              "timeout": 300,
            }
          }
          onClosed={[Function]}
          onOpened={[Function]}
          role="dialog"
          toggle={[Function]}
          zIndex={1050}
        >
          <ModalHeader
            charCode={215}
            closeAriaLabel="Close"
            tag="h5"
            toggle={[Function]}
            wrapTag="div"
          >
            Filter
          </ModalHeader>
          <ModalBody
            tag="div"
          >
            <div
              className="search-filters"
            >
              <FilterTree
                activeFilter={null}
                aggregations={null}
                collapseItems={false}
                handleFiltering={[Function]}
                referenceDataItems={null}
                title="Provider"
              />
              <FilterBox
                activeFilter={null}
                filter={
                  Object {
                    "buckets": Array [
                      Object {
                        "count": 1,
                        "key": "MISSING",
                      },
                      Object {
                        "count": 1,
                        "key": "application/json",
                      },
                      Object {
                        "count": 1,
                        "key": "application/xml",
                      },
                      Object {
                        "count": 1,
                        "key": "text/json",
                      },
                      Object {
                        "count": 1,
                        "key": "text/xml",
                      },
                    ],
                  }
                }
                htmlKey={2}
                onClick={[Function]}
                themesItems={null}
                title="Format"
              />
            </div>
          </ModalBody>
          <ModalFooter
            tag="div"
          >
            <Button
              className="fdk-button"
              color="primary"
              onClick={[Function]}
              tag="button"
            >
              Close
            </Button>
          </ModalFooter>
        </Modal>
        <FilterTree
          activeFilter={null}
          aggregations={null}
          collapseItems={false}
          handleFiltering={[Function]}
          referenceDataItems={null}
          title="Provider"
        />
        <FilterBox
          activeFilter={null}
          filter={
            Object {
              "buckets": Array [
                Object {
                  "count": 1,
                  "key": "MISSING",
                },
                Object {
                  "count": 1,
                  "key": "application/json",
                },
                Object {
                  "count": 1,
                  "key": "application/xml",
                },
                Object {
                  "count": 1,
                  "key": "text/json",
                },
                Object {
                  "count": 1,
                  "key": "text/xml",
                },
              ],
            }
          }
          htmlKey={2}
          onClick={[Function]}
          themesItems={null}
          title="Format"
        />
      </div>
    </aside>
    <div
      className="col-12 col-lg-8"
      id="apis"
    >
      <SearchHitItem
        fadeInCounter={0}
        item={
          Object {
            "formats": Array [],
            "id": "68dda78f-364f-41dd-869a-0dd8d9a3fe4c",
            "nationalComponent": false,
            "publisher": Object {
              "id": "970018131",
              "name": "UTDANNINGSDIREKTORATET",
              "orgPath": "/STAT/872417842/970018131",
              "prefLabel": Object {
                "no": "Utdanningsdirektoratet",
              },
            },
            "title": "Barnehagefakta",
            "titleFormatted": "Barnehagefakta",
          }
        }
        key="68dda78f-364f-41dd-869a-0dd8d9a3fe4c"
        publishers={null}
        referenceData={null}
      />
      <SearchHitItem
        fadeInCounter={1}
        item={
          Object {
            "formats": Array [
              "application/xml",
              "application/json",
              "text/json",
              "text/xml",
            ],
            "id": "70f6e5b3-b633-413b-b5e7-11b63bb413eb",
            "nationalComponent": false,
            "publisher": Object {
              "id": "970018131",
              "name": "UTDANNINGSDIREKTORATET",
              "orgPath": "/STAT/872417842/970018131",
              "prefLabel": Object {
                "no": "Utdanningsdirektoratet",
              },
            },
            "title": "Nasjonalt barnehageregister API",
            "titleFormatted": "Nasjonalt barnehageregister API",
          }
        }
        key="70f6e5b3-b633-413b-b5e7-11b63bb413eb"
        publishers={null}
        referenceData={null}
      />
      <div
        className="col-12 d-flex justify-content-center"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Sidepaginering.
        </span>
        <PaginationBoxView
          activeClassName="active"
          breakClassName="break-me"
          breakLabel={
            <span>
              ...
            </span>
          }
          containerClassName="pagination"
          disableInitialCallback={true}
          disabledClassName="disabled"
          forcePage={0}
          marginPagesDisplayed={1}
          nextClassName="next"
          nextLabel="Neste side"
          onPageChange={[Function]}
          pageCount={1}
          pageRangeDisplayed={2}
          previousClassName="previous"
          previousLabel="Forrige side"
          subContainerClassName="pages pagination"
        />
      </div>
    </div>
  </div>
</main>
`;

exports[`should render ResultsApi correctly with minimum of props and no hits 1`] = `
<main
  data-test-id="apis"
  id="content"
>
  <div
    className="row mb-3"
  >
    <div
      className="col-6 col-lg-4"
    />
    <div
      className="col-6 col-lg-4 offset-lg-4"
    >
      <div
        className="d-flex justify-content-end"
      >
        <Button
          className="fdk-button fdk-button-black-toggle selected"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Relevance
        </Button>
        <Button
          className="fdk-button fdk-button-black-toggle"
          color="primary"
          onClick={[Function]}
          tag="button"
        >
          Recently published
        </Button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <aside
      className="search-filters col-lg-4 d-none d-lg-block"
    >
      <span
        aria-hidden="false"
        className="uu-invisible"
      >
        Filtrering
      </span>
      <FilterPills
        history={
          Object {
            "push": [Function],
          }
        }
        location={
          Object {
            "search": "",
          }
        }
        locationSearch={Object {}}
        losItems={null}
        publishers={null}
        themesItems={null}
      />
    </aside>
    <div
      className="col-12 col-lg-8"
      id="apis"
    >
      <div
        className="col-12 d-flex justify-content-center"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Sidepaginering.
        </span>
        <PaginationBoxView
          activeClassName="active"
          breakClassName="break-me"
          breakLabel={
            <span>
              ...
            </span>
          }
          containerClassName="pagination"
          disableInitialCallback={true}
          disabledClassName="disabled"
          forcePage={0}
          marginPagesDisplayed={1}
          nextClassName="next"
          nextLabel="Neste side"
          onPageChange={[Function]}
          pageCount={1}
          pageRangeDisplayed={2}
          previousClassName="previous"
          previousLabel="Forrige side"
          subContainerClassName="pages pagination"
        />
      </div>
    </div>
  </div>
</main>
`;
